### PR TITLE
Hyperparams tuned; plots collected

### PR DESCRIPTION
**Pull request by d-senyaka**
Experiments for BEFORE & BEFOREAFTER; grid d_model×layers; attention plots.

### Milestone Summary

- All experiments (BEFORE & BEFOREAFTER) complete.
- Achieved > 99 % accuracy with d64 L2 model.
- Plots confirm progressive attention depth effects:

  1. L1 → diagonal (local)
  2. L2 → multi-hop global
  3. L4 → diffuse redundant.

<html>
<body>
<h3 data-start="265" data-end="295"><strong data-start="269" data-end="295">Acceptance Checklist</strong></h3>
<div class="_tableContainer_1rjym_1"><div tabindex="-1" class="group _tableWrapper_1rjym_13 flex w-fit flex-col-reverse">

Requirement | Status | Notes
-- | -- | --
≥ 1 run per configuration (L1, L2, L4, d64/d128, BEFORE/BEFOREAFTER) | ✅ | All 6 major configs executed successfully
At least 1 tuned config ≥ 95 % dev accuracy on BEFORE | ✅ | d64 / L2 / lr = 1e-3 / ep = 5 → 99.53 %
Attention plots collected for 1/2/4 layers | ✅ | Stored under experiments/plots__task-BEFORE__d64__L{1,2,4}
BEFOREAFTER task runs validated (≥ 98 %) | ✅ | Confirms robust causal tracking
experiments/RESULTS.md populated with metrics | ✅ | Logged for traceability
Final config file noted (final_config_part1.txt) | ✅ | Lightest stable model recorded
All runs reproducible in current venv | ✅ | Env: Python 3.11 + torch 2.9 + numpy 2.3 + matplotlib 3.10

</div></div>
</body>
</html>

<html>
<body>
<h3 data-start="1078" data-end="1115"><strong data-start="1082" data-end="1115">Deliverables in Repository</strong></h3>
<div class="_tableContainer_1rjym_1"><div tabindex="-1" class="group _tableWrapper_1rjym_13 flex w-fit flex-col-reverse">

File / Directory | Description
-- | --
part_1_encoder/transformer.py | Final Transformer implementation + training loop
experiments/RESULTS.md | Markdown log of all experiments
experiments/final_config_part1.txt | Final tuned model parameters
experiments/model__*.pt | Saved PyTorch checkpoints
experiments/meta__*.json | Metadata and accuracy history

</div></div>
</body>
</html>

# Part 1 — Transformer Encoder Experiments Log

| Task | d_model | Layers | LR | Epochs | Dev Acc | Train Acc | Tag |
|------|:--------:|:------:|:--:|:-------:|:--------:|:----------:|:----|
| BEFORE | 64 | 1 | 1e-3 | 5 | **97.26%** | 97.25% | task-BEFORE__d64__L1__lr1e-3__ep5 |
| BEFORE | 64 | 2 | 1e-3 | 5 | **99.53% ✅** | 99.20% | task-BEFORE__d64__L2__lr1e-3__ep5 |
| BEFORE | 64 | 4 | 1e-3 | 5 | 98.30% | 97.75% | task-BEFORE__d64__L4__lr1e-3__ep5 |
| BEFORE | 128 | 2 | 8e-4 | 8 | 90.39% | 89.85% | task-BEFORE__d128__L2__din128__lr8e-4__ep8 |
| BEFOREAFTER | 64 | 1 | 1e-3 | 5 | 79.19% | 79.00% | task-BEFOREAFTER__d64__L1__lr1e-3__ep5 |
| BEFOREAFTER | 64 | 2 | 1e-3 | 5 | **98.65% ✅** | 98.50% | task-BEFOREAFTER__d64__L2__lr1e-3__ep5 |

---

### 🧠 Summary Notes

- **Best configuration:** `d64 / L2 / lr=1e-3 / epochs=5`
- Achieves **99.5%** on *BEFORE* and **98.6%** on *BEFOREAFTER*.
- Models with ≥2 layers exhibit stronger global reasoning via multi-step attention.
- Wider or deeper variants (d128, L4) show diminishing returns or instability.
- This file serves as the lightweight log for replication or cross-member verification.

Final Tuned Configuration – Part 1 (Encoder)
--------------------------------------------
Task: BEFORE
Model: Transformer(d_model=64, num_layers=2, d_internal=64)
Learning Rate: 1e-3
Epochs: 5
Dev Accuracy: 99.53%
Checkpoint: experiments/model__task-BEFORE__d64__L2__lr1e-3__ep5.pt
Justification: Lightest model exceeding 95% with consistent stability.

### CODE REVIEW

**Comment by d-senyaka** [comment]
<html>
<body>
<h3 data-start="1729" data-end="1764"><strong data-start="1733" data-end="1764">Final Tuned Configuration</strong></h3>
<div class="_tableContainer_1rjym_1"><div tabindex="-1" class="group _tableWrapper_1rjym_13 flex w-fit flex-col-reverse">

Parameter | Final Value
-- | --
Task | BEFORE (normal task)
d_model | 64
Layers | 2
d_internal | 64
LR | 1e-3
Epochs | 5
Dev Accuracy | 99.53 %
Checkpoint | experiments/model__task-BEFORE__d64__L2__lr1e-3__ep5.pt
Metadata | experiments/meta__task-BEFORE__d64__L2__lr1e-3__ep5.json

</div></div>
</body>
</html>
